Source/Urho3D/CSharp/Managed/Container/SharedPtr.cs

```diff
@@ -0,0 +1,146 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Urho3DNet
+{
+    public class SharedPtr<T> : IDisposable, IEquatable<SharedPtr<T>> where T : RefCounted
+    {
+        /// <summary>
+        /// Construct immutable SharedPtr object.
+        /// Don't forget to dispose it.
+        /// </summary>
+        /// <param name="other">Object to store pointer to.</param>
+        public SharedPtr(T other)
+        {
+            if (other?.IsNotExpired == true)
+            {
+                _value = other;
+                _value.AddRef();
+            }
+            else
+                _value = null;
+        }
+
+        /// <summary>
+        /// SharedPtr finalizer.
+        /// Should not be executed on not-disposed SharedPtr.
+        /// </summary>
+        ~SharedPtr()
+        {
+            if (_value?.IsNotExpired == true)
+                Trace.WriteLine(
+                    $"Object of type {typeof(T).FullName} wasn't properly disposed. Please call Dispose on all SharedPtr<{typeof(T).Name}> instances.");
+        }
+
+        /// <summary>
+        /// Copy SharedPtr.
+        /// </summary>
+        /// <param name="other">SharedPtr instance to copy</param>
+        public SharedPtr(SharedPtr<T> other) : this(other?._value)
+        {
+        }
+
+        /// <summary>Indicates whether the current object is equal to another object of the same type.</summary>
+        /// <param name="other">An object to compare with this object.</param>
+        /// <returns>true if the current object is equal to the <paramref name="other">other</paramref> parameter; otherwise, false.</returns>
+        public bool Equals(SharedPtr<T> other)
+        {
+            if (ReferenceEquals(null, other)) return false;
+            if (ReferenceEquals(this, other)) return true;
+            return EqualityComparer<T>.Default.Equals(_value, other._value);
```

Comment:
I'm not sure this is 100% correct. Wrapper types do not define `Equals()` most of the time, and when they do - they forward equality comparison to C++ `==` operator. In contrast, C++ `SharedPtr<>` does instance equality. I think the right thing to do here is to do `T.getCPtr(_value).Handle == T.getCPtr(other._value).Handle` which will compare pointers to C++ instances.

Reply:
No reason to waste time on equality methods now if they aren't in use. I'm going to delete this.